keychain_win.cpp

```diff
@@ -0,0 +1,211 @@
+/*
+ * Copyright (c) 2013 GitHub Inc.
+ * Copyright (c) 2019 Hannes Rantzsch
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining
+ * a copy of this software and associated documentation files (the
+ * "Software"), to deal in the Software without restriction, including
+ * without limitation the rights to use, copy, modify, merge, publish,
+ * distribute, sublicense, and/or sell copies of the Software, and to
+ * permit persons to whom the Software is furnished to do so, subject to
+ * the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+ * LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+ * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ *
+ */
+
+// clang-format off
+// make sure windows.h is included before wincred.h
+#include "keychain.h"
+
+#include <memory>
+
+#define UNICODE
+
+#include <windows.h>
+#include <wincred.h>
+#include <intsafe.h> // for DWORD_MAX
+// clang-format on
+
+namespace {
+
+static DWORD cred_type = CRED_TYPE_GENERIC;
+
+struct LpwstrDeleter {
+    void operator()(WCHAR *p) const { delete[] p; }
+};
+using ScopedLpwstr = std::unique_ptr<WCHAR, LpwstrDeleter>;
+
+LPWSTR utf8ToWideChar(const std::string &utf8) {
+    int len = MultiByteToWideChar(
+        CP_UTF8,
+        0, // flags must be 0 for UTF-8
+        utf8.c_str(),
+        -1,      // rely on null-terminated input string
+        nullptr, // no out-buffer needed
+        0); // return required buffer size rather than writing to an out-buffer
+    if (len == 0) {
+        return nullptr;
+    }
```

Comment:
No, for an empty string it will indicate a required buffer size of 1, because it includes the terminating null character. I've improved the naming and added a comment.